### ISSUE TITLE
[BUG] High memory usage in 1.7.0rc1 during consumption compared to v1.6.0

### ISSUE THREAD

**Issue by Rondal** [paperless-ngx/paperless-ngx#768]
### Description

The paperless webserver container is consuming way more memory than with v1.6.0 when consuming a document. Configuration as follows:

Proxmox VM Host, Core i5 (4C/8T), 32G Memory
-> On that: VM with 4 cores, 4GB of memory / Ubuntu 20.04.04 LTS
-> On that: Docker installation of paperless-ngx 1.7.0rc1

Configured with 2 workers and 2 threads. Worked perfectly fine with 1.6.0. With 1.7.0rc1 I noticed heavy swapping. When giving it 8G of memory the swapping was resolved but barely. Still topping out at 7.8G of total memory consumption (Ubuntu included -> idle memory consumption of whole system is ~800-900M).

After consumption it goes back to normal. As far as I understand the process it is happening in the OCRMyPDF phase of consumption. Propably during unpaper or tesseract. Setting PAPERLESS_CONVERT_MEMORY_LIMIT does not have any effect on the issue.

Downgrade to 1.6.0 fixes it. I verified the issue on a different host with a different OS (Docker on Alpine on other VM host). Same issue. Each Thread of a worker consumes around 1.5G. See attached screenshots below for involved processes.

I'm unqualified to look into the processes in more detail and am looking for confirmation and/or guidance how to troubleshoot. Of course I'm happy to help in any way. 

### Expected behavior

I'd expect the memory consumption not to differ that dramatically from v1.6.0 (1.5G vs well below 1G). Or some option or mechanism to honor the physical memory limits.

### Steps to reproduce

1. Install 1.6.0
2. Consume document and note max memory consumption
3. Upgrade (or reinstall) 1.7.0
4. Consume same document and compare max memory consumption

### Webserver logs

```bash
n/a - No errors.
```

### Screenshots

Under load:
![load](https://user-images.githubusercontent.com/217929/164068233-6980a2a1-7405-4932-baa6-1a1685792720.png)

Versus idle:
![idle](https://user-images.githubusercontent.com/217929/164068247-41796d25-92a9-4adf-b203-98d06c627d04.png)

### Paperless-ngx version

1.7.0rc1

### Host OS

Ubuntu 20.04.04 LTS

### Installation method

Docker

**Comment by qcasey** [comment]
Thanks for reporting. That's an immense amount of memory. Can you describe your document setup a little for us? I'll do my best to replicate it:

```
Approximate # of documents total:
Approximate # of documents consuming (when issue occurs):
Extra languages enabled:
IMAP / Mail account enabled:
```

For me, `beta` idles and consumes documents with the same memory usage as `1.6.0`. That's ~380mb and ~480mb respectively.

**Comment by stumpylog** [comment]
If I really, really stress out the beta image (dumping ~400 documents into the consume folder), I will see the memory usage spike to very high levels.

However, I will also see the memory start to decrease as document consumption starts finishing (or maybe garbage collection happens), here dropping from 32GB to about 11GB:
![image](https://user-images.githubusercontent.com/797416/164129334-a52271ac-e437-4d8d-82ae-b11a55cfe696.png)

Then dropping again to ~440 MB after completing.

**Comment by shamoon** [comment]
Yea I also cant reproduce this, both 1.6.0 and 1.7.0-rc1 hovering around 300mb for me

<img width="645" alt="Screen Shot 2022-04-19 at 7 33 10 PM" src="https://user-images.githubusercontent.com/4887959/164135493-66ed6f45-62da-4d86-987c-dd5ddf08a899.png">

**Comment by Rondal** [comment]
As for the document setup: Nothing fancy. I did all testing with the same set of docs. 3 PDF files, 1 to 4 pages, put into consumption dir. 

Scanned with a brother ADS-2600W, color, 600dpi. All docs combined are about 9M.

**Comment by Rondal** [comment]
I created two test documents which I can share. Both are of the same content and scanned with the same settings as the document I used for testing yesterday. They are used with the new barcode splitting feature but can also be used without it. The problem I described appears both ways.

1 page, separator page, 2 pages, separator page, 3 pages, separator page, 4 pages

Any more pages make no more difference as memory usages directly relates to threads which are limited to 4 in my case. Giving it more or less threads scales the consumption in a linear manner (1.5G per thread).

One 300dpi version, one 600dpi version (as used for testing). With the 600dpi version and 4 threads the memory consumption peaks at about 1.5G per thread. 300dpi is way lower at about 300M-500M per thread. Still higher than 1.6.0.

Files attached
[testfile_300dpi.pdf](https://github.com/paperless-ngx/paperless-ngx/files/8519562/testfile_300dpi.pdf)
[testfile_600dpi.pdf](https://github.com/paperless-ngx/paperless-ngx/files/8519565/testfile_600dpi.pdf)
.

**Comment by Rondal** [comment]
I just noticed that I missed commenting on two of the questions.

Consumption is by putting the files into the consumption dir, which in this case is on a CIFS share but the issue also occurs with a local consumption dir. OCR language is German. Complete configuration below.

Also just to clarify: I'm absolutely not complaining about the memory the system uses when idle. That's in line with 1.6.0. It is just the memory usage during consumption process. Of course I can just give the system more memory but the difference between 1.6.0 and 1.7.0rc1 gives me pause and made me think that something might be off.

docker-compose.yml:

> version: "3.4"
> services:
>   broker:
>     image: redis:6.0
>     restart: unless-stopped
>     volumes:
>       - redisdata:/data
> 
>   db:
>     image: postgres:13
>     restart: unless-stopped
>     volumes:
>       - /home/dms/paperless-ngx/db:/var/lib/postgresql/data
>     environment:
>       POSTGRES_DB: paperless
>       POSTGRES_USER: paperless
>       POSTGRES_PASSWORD: paperless
> 
>   webserver:
>     image: ghcr.io/paperless-ngx/paperless-ngx:beta
>     restart: unless-stopped
>     depends_on:
>       - db
>       - broker
>       - gotenberg
>       - tika
>     ports:
>       - 80:8000
>     healthcheck:
>       test: ["CMD", "curl", "-f", "http://localhost:8000"]
>       interval: 30s
>       timeout: 10s
>       retries: 5
>     volumes:
>       - /home/dms/paperless-ngx/data:/usr/src/paperless/data
>       - /media/dms:/usr/src/paperless/media
>       - ./export:/usr/src/paperless/export
>       - /media/dms/Inbox/Consumer:/usr/src/paperless/consume
>     env_file: docker-compose.env
>     environment:
>       PAPERLESS_REDIS: redis://broker:6379
>       PAPERLESS_DBHOST: db
>       PAPERLESS_TIKA_ENABLED: 1
>       PAPERLESS_TIKA_GOTENBERG_ENDPOINT: http://gotenberg:3000
>       PAPERLESS_TIKA_ENDPOINT: http://tika:9998
> 
>   gotenberg:
>     image: gotenberg/gotenberg:7
>     restart: unless-stopped
>     environment:
>       CHROMIUM_DISABLE_ROUTES: 1
> 
>   tika:
>     image: apache/tika
>     restart: unless-stopped
> 
> volumes:
>   data:
>   media:
>   pgdata:
>   redisdata:
> 

docker-compose.env:

> USERMAP_UID=1001
> USERMAP_GID=1001
> PAPERLESS_TIME_ZONE=Europe/Berlin
> PAPERLESS_OCR_LANGUAGE=deu
> PAPERLESS_SECRET_KEY=xxx
> PAPERLESS_CONSUMER_RECURSIVE=true
> PAPERLESS_CONSUMER_SUBDIRS_AS_TAGS=true
> PAPERLESS_CONSUMER_POLLING=5
> PAPERLESS_FILENAME_DATE_ORDER=YMD
> PAPERLESS_CONSUMER_IGNORE_PATTERNS=[".DS_STORE/*", "._*", ".stfolder/*", "@eaDir/*"]
> PAPERLESS_FILENAME_FORMAT={document_type}/{correspondent}/{created_year}{created_month}{created_day}-{title}-{tag_list}
> PAPERLESS_TASK_WORKERS=1
> PAPERLESS_THREADS_PER_WORKER=4
> PAPERLESS_CONSUMER_DELETE_DUPLICATES=true
> PAPERLESS_OCR_ROTATE_PAGES_THRESHOLD=9
> PAPERLESS_CONSUMER_ENABLE_BARCODES=true
> PAPERLESS_CONSUMER_BARCODE_STRING=ADAR-NEXTDOC

**Comment by smseidl** [comment]
My RPI is about the same size.  I will try to do some testing over the next couple days (scanner is getting replaced again 👎 ) and see what I have happen.  @stumpylog what did you use to create your graph?

**Comment by shamoon** [comment]
My bad for not noting the only during consumption part. Mines actually even more greedy, but not really a difference between the two versions... both spike > 4gb and then come back down.

<img width="997" alt="Screen Shot 2022-04-20 at 7 08 50 AM" src="https://user-images.githubusercontent.com/4887959/164249698-4f3c0072-c424-40ff-85ef-537e80a942e7.png">

**Comment by stumpylog** [comment]
I was pulling the usage from netdata, which probably is grabbing either from Docker or maybe cgroups.  It's also most likely the virtual memory, not resident memory

**Comment by Rondal** [comment]
Regarding @shamoon observation with no difference between the versions: I tested again and recorded some data. I do see a clear difference with both versions of the above files (300dpi and 600dpi). 

I‘ll graph and upload the data to tonight.

**Comment by shamoon** [comment]
Sorry, I didnt mean to imply youre not seeing that, just that its not happening for me. Im also testing on Ubuntu 20.04 LTS under docker, FWIW.

Im not sure if this was mentioned (here or on matrix) but the only significant feature added to 1.7.0 to the consumption process is #532 (unless someone else correct me if Im wrong?), which is disabled by default. So as mentioned this, could very well be in one of our dependencies, even something OS-specific.

I guess the underlying question is whether there really is a bug here (/ where), or rather if we feel the resources usage is unacceptable when consuming complex high dpi files like the test files. Even my 1.6.0 is slow with these test files as opposed to more typical files I throw at it.

**Comment by qcasey** [comment]
> Files attached
[testfile_300dpi.pdf](https://github.com/paperless-ngx/paperless-ngx/files/8519562/testfile_300dpi.pdf)
[testfile_600dpi.pdf](https://github.com/paperless-ngx/paperless-ngx/files/8519565/testfile_600dpi.pdf)

These test files are great, I see much higher spikes of memory using those with `beta`.

Ubuntu 20 LTS, 4G RAM:
![Screenshot from 2022-04-20 11-13-06](https://user-images.githubusercontent.com/1429672/164295737-e79264b8-2da0-40d9-a0be-222ba3c232a3.png)


It seems the only tuning factor we currently have is `PAPERLESS_WEBSERVER_WORKERS`, which unfortunately is fairly blunt.

> I guess the underlying question is whether there really is a bug here (/ where), or rather if we feel the resources usage is unacceptable when consuming complex high dpi files like the test files. Even my 1.6.0 is slow with these test files as opposed to more typical files I throw at it.

Perhaps there is a (relatively) easy configuration option we missed in a dependency upgrade which could tune this down. I don't see this behavior on any of the PDFs I normally use for testing, including some 50+MB in size. It's difficult to determine how much of a problem this is.

**Comment by stumpylog** [comment]
Using this command: `/usr/bin/time -v ocrmypdf --deskew --rotate-pages --clean --skip-text testfile_600dpi.pdf temp.pdf` (roughly the equivalent of our defaults)

| ocrmypdf version | max resident set size |
| -------- | --------- |
|  13.4.2 | 5.34 GB |
| 13.4.0 |  5.37 GB |

So we didn't step this up very much in the beta and it doesn't look to have any real effect.  It could always be tesseract, unpaper or ghostscript versions though.

**Comment by Rondal** [comment]
> > I guess the underlying question is whether there really is a bug here (/ where), or rather if we feel the resources usage is unacceptable when consuming complex high dpi files like the test files. Even my 1.6.0 is slow with these test files as opposed to more typical files I throw at it.
> 
> Perhaps there is a (relatively) easy configuration option we missed in a dependency upgrade which could tune this down. I don't see this behavior on any of the PDFs I normally use for testing, including some 50+MB in size. It's difficult to determine how much of a problem this is.

As I can reproduce this basically with each and every scan of my ADS2800W I was thinking that it might be specific to the way this machine creates PDF files. As I did not have any strange behaviour in any other app I would still say there might be something strange happening in the dependencies. Anyway: I'll try some different creation settings tomorrow and also create some PDFs for testing that do not come from the scanner just to see if the results might tell something.

**Comment by shamoon** [comment]
Definitely appreciate all the investigating! Yes let us know if you can pin it down.

**Comment by Rondal** [comment]
All right. Did some more testing with my scanner and its PDF files. There are basically only two options which might have an effect on the created PDF which are document size detection and 'file size' (basically compression). I tested different variations of that without any really interesting results. I'll attach the results and test files below.

I couldn't really measure memory consumption when feeding a digitally created document because consumption was finished within seconds without any real peaks in CPU or Memory.

Chart:
<img width="950" alt="Screenshot 2022-04-21 at 13 29 39" src="https://user-images.githubusercontent.com/217929/164449477-523ec8e8-0dbe-484d-a8aa-94952e6e6a4e.png">

Testfiles:
[4 pages, 600dpi, auto, medium.pdf](https://github.com/paperless-ngx/paperless-ngx/files/8530563/4.pages.600dpi.auto.medium.pdf)

[4 pages, 600dpi, a4, medium.pdf](https://github.com/paperless-ngx/paperless-ngx/files/8530560/4.pages.600dpi.a4.medium.pdf)

[4 pages, 600dpi, a4, large.pdf](https://github.com/paperless-ngx/paperless-ngx/files/8530559/4.pages.600dpi.a4.large.pdf)

[4 pages, 600dpi, a4, small.pdf](https://github.com/paperless-ngx/paperless-ngx/files/8530562/4.pages.600dpi.a4.small.pdf)

**Comment by Rondal** [comment]
Did a bit more debugging. I still can reproduce the difference between 1.6.0 and 1.7.0rc1 but it seems that my specific PDF files (generated with said scanner) are in some kind special. When using different PDFs, even with 600DPI, I can still see the difference between the versions, but the overall memory consumption is so much lower that it hardly matters at all.

Also I found the specific moment in processing where the memory consumption goes up to the seen levels:

```
    4 Running: ['tesseract', '-l', 'eng', '--psm', '2', '/tmp/ocrmypdf.io.vbcj993e/000004_rasterize.png', 'stdout']
    2 Running: ['tesseract', '-l', 'eng', '--psm', '2', '/tmp/ocrmypdf.io.vbcj993e/000002_rasterize.png', 'stdout']
    3 Running: ['tesseract', '-l', 'eng', '--psm', '2', '/tmp/ocrmypdf.io.vbcj993e/000003_rasterize.png', 'stdout']
    1 Running: ['tesseract', '-l', 'eng', '--psm', '2', '/tmp/ocrmypdf.io.vbcj993e/000001_rasterize.png', 'stdout']
    4 Running: ['tesseract', '-l', 'eng', '--psm', '2', '/tmp/ocrmypdf.io.vbcj993e/000004_rasterize.png', 'stdout']
    2 Running: ['tesseract', '-l', 'eng', '--psm', '2', '/tmp/ocrmypdf.io.vbcj993e/000002_rasterize.png', 'stdout']
    3 Running: ['tesseract', '-l', 'eng', '--psm', '2', '/tmp/ocrmypdf.io.vbcj993e/000003_rasterize.png', 'stdout']
    1 Running: ['tesseract', '-l', 'eng', '--psm', '2', '/tmp/ocrmypdf.io.vbcj993e/000001_rasterize.png', 'stdout']
    4 Running: ['unpaper', '-v', '--dpi', '600.34965', '--layout', 'none', '--mask-scan-size', '100', '--no-border-align', '--no-mask-center', '--no-grayfilter', '--no-blackfilter', '--no-deskew', '/tmp/tmpy629is3s/input.pnm', '/tmp/tmpy629is3s/output.ppm']
    2 Running: ['unpaper', '-v', '--dpi', '600.350263', '--layout', 'none', '--mask-scan-size', '100', '--no-border-align', '--no-mask-center', '--no-grayfilter', '--no-blackfilter', '--no-deskew', '/tmp/tmpvq0c3vb4/input.pnm', '/tmp/tmpvq0c3vb4/output.ppm']
    1 Running: ['unpaper', '-v', '--dpi', '600.174825', '--layout', 'none', '--mask-scan-size', '100', '--no-border-align', '--no-mask-center', '--no-grayfilter', '--no-blackfilter', '--no-deskew', '/tmp/tmpu4_amaf6/input.pnm', '/tmp/tmpu4_amaf6/output.ppm']
    3 Running: ['unpaper', '-v', '--dpi', '600.175131', '--layout', 'none', '--mask-scan-size', '100', '--no-border-align', '--no-mask-center', '--no-grayfilter', '--no-blackfilter', '--no-deskew', '/tmp/tmp0oz381x1/input.pnm', '/tmp/tmp0oz381x1/output.ppm']
    4 stdout/stderr = [ppm_pipe @ 0x564889ee02c0] Stream #0: not enough frames to estimate rate; consider increasing probesize
```
Thats the part of ocrmypdf after rotation and directly before it invokes unpaper. Memory consumption stays high until after unpaper is done.

So I actually have no idea where to go with this from here. As this seems to be unrelated to paperless I'd say we can close off this issue here. On the other hand that specific scanner model is not that rare out there and as paperless is very much built for an environment where you scan your stuff, this might affect more users who are then wondering why their Raspi kills random processes.

Maybe someone has an idea what is so special about those PDFs? Might be worth a note in the documentation? Let me know what you think :)

**Comment by qcasey** [comment]
Thank you for investigating this further; the chart, test files, and running processes are very useful! 

We have a couple avenues we can take here, I think the broadest/most comprehensive being "Add function to limit paperless' memory consumption". Paperless shouldn't put itself in a place to be OOM killed regardless of the input. Another option (not mutually exclusive) is to wrangle our dependencies to consume less in certain scenarios. Another idea is to rebound from a killed subprocess and notify the user, if that's even possible.

However all of this is a tall order for an otherwise niche issue. I don't personally have the skillset to help. I think we should keep this issue open (as it is still a problem), but I expect a direct bug-fix is a long way off.

Please update us if you have any other useful findings as someone might take a look at this much later. Again, really appreciate the deep dive you did here :)

> Chart: <img alt="Screenshot 2022-04-21 at 13 29 39" width="950" src="https://user-images.githubusercontent.com/217929/164449477-523ec8e8-0dbe-484d-a8aa-94952e6e6a4e.png">
> 
> Testfiles: [4 pages, 600dpi, auto, medium.pdf](https://github.com/paperless-ngx/paperless-ngx/files/8530563/4.pages.600dpi.auto.medium.pdf)
> 
> [4 pages, 600dpi, a4, medium.pdf](https://github.com/paperless-ngx/paperless-ngx/files/8530560/4.pages.600dpi.a4.medium.pdf)
> 
> [4 pages, 600dpi, a4, large.pdf](https://github.com/paperless-ngx/paperless-ngx/files/8530559/4.pages.600dpi.a4.large.pdf)
> 
> [4 pages, 600dpi, a4, small.pdf](https://github.com/paperless-ngx/paperless-ngx/files/8530562/4.pages.600dpi.a4.small.pdf)

**Comment by panther-sh** [comment]
I'am also affected by the problem. 
Since the Update from 1.6.0 -> 1.7.0 the raspberry pi runs out of memory while it consume some documents and then seems to be freezed. 
In the last few weeks the 1.6.0 version proccessed about 800 documents without any out of memory problems. This also occurs when PAPERLESS_TASK_WORKER and PAPERLESS_THREADS_PER_WORKER set to 1.
At Version 1.6.0 it works with PAPERLESS_TASK_WORKER and PAPERLESS_THREADS_PER_WORKER set to 2 without out of memory problems.

Operating conditions:
- Raspberry Pi 4 / 2GB Ram
- Raspberry Pi OS 64bit 
- Docker version 20.10.14
- Paperless-ngx 1.7.0

**Comment by Rondal** [comment]
That was basically what I was afraid of but not able to test. On a Raspi you're just not able to workaround the problem with maxing out on memory.

@panther-sh: As this seems to affect only specific PDF. How did you create the files which result in this issue?

**Comment by Rondal** [comment]
> We have a couple avenues we can take here, I think the broadest/most comprehensive being "Add function to limit paperless' memory consumption". **Paperless shouldn't put itself in a place to be OOM killed regardless of the input.** Another option (not mutually exclusive) is to wrangle our dependencies to consume less in certain scenarios. Another idea is to rebound from a killed subprocess and notify the user, if that's even possible.

Emphasis by me. Problem I see is that Paperless itself is not the offending software. It seems to be ocrmypdf, unpaper or tesseract (or a combination). But I don't understand the interworkings of all tools well enough to raise a bug over there. Question is if it is even possible to do something from the Paperless side. At a first glance I couldn't find any config switches for the mentioned tools which would enable a memory limit.

**Comment by stumpylog** [comment]
One possible (temporary) fix is setting a memory limit for the container: https://docs.docker.com/compose/compose-file/compose-file-v3/#resources

I'm unsure exactly how this may affect speed of processing, but it will likely slow it down, while still allowing the service to run.  On a Pi, it may swap to disk a lot

I checked what I image are the relevant tools between 1.6 and 1.7:
| tool | version 1.6 | version 1.7 |
| -------- | --------- | ------------ |
|  ocrmypdf  | 13.4.0  |  13.4.2 |
| tesseract | 4.1.1  | 4.1.1 |
| unpaper | 6.1  |  6.1 |
| gs (ghostscript)| 9.53.3  | 9.53.3 |
| pdfminer.six | 2021.10.12  | 2022.03.19 |

So basically no version changes, only a minor bump to ocrmypdf, and just 1 version in pdfminer.six.  I do find [pdfminer.six/issues/723](https://github.com/pdfminer/pdfminer.six/issues/723), which could be related?

**Comment by smseidl** [comment]
Raspberry Pi OS 64bit - this hasn't been officially out for all that long right?  6 months ago, I went with a Ubuntu 64bit cause I couldn't find a Raspberry Pi OS 64bit.  I'm not sure if we have many running that for testing.  @panther-sh are you running with or without the Barcode Splitter enabled?

I haven't processed a huge number of docs since upgrading to 1.7.0 on my RPI4 /4 GB Ram running Ubuntu 20.04.4 LTS but I haven't noticed this problem.

**Comment by panther-sh** [comment]
@Rondal The files are created with an Epson WorkeForce Pro WF-C5710 scanner. The file format ist .pdf with 300 DPI.
@smseidl The barcode splitter ist disabled. I will test version 1.7.0 with Ubuntu 22.04 LTS in the next day!

The memory limit for docker containers is also a good idea. I'll give it a try and take a look at the processing speed.

**Comment by Rondal** [comment]
I'm not a docker or kernel specialist, but I'm not sure if limiting the container would help:

First of all, a limited containers application might not be aware of the limit, depending where it looks for the available memory and of course if it cares at all:
```
dock1:~# docker run --memory 100m --memory-swap 100m --rm busybox free -m
              total        used        free      shared  buff/cache   available
Mem:           3936         560        1105          21        2270        3121
Swap:          3976         112        3864
dock1:~# docker run --memory 100m --memory-swap 100m --rm busybox cat /sys/fs/cgroup/memory/memory.limit_in_bytes
104857600
```
Second a limited docker container, to my knowledge, just kills processes if they try to consume more than allowed:
```
dock1:~# docker run --memory 50m --memory-swap 50m --rm -it progrium/stress --vm 1 --vm-bytes 62914560 --timeout 1s
stress: info: [1] dispatching hogs: 0 cpu, 0 io, 1 vm, 0 hdd
stress: dbug: [1] using backoff sleep of 3000us
stress: dbug: [1] setting timeout to 1s
stress: dbug: [1] --> hogvm worker 1 [7] forked
stress: dbug: [7] allocating 62914560 bytes ...
stress: dbug: [7] touching bytes in strides of 4096 bytes ...
stress: FAIL: [1] (416) <-- worker 7 got signal 9
stress: WARN: [1] (418) now reaping child worker processes
stress: FAIL: [1] (422) kill error: No such process
stress: FAIL: [1] (452) failed run completed in 0s
dock1:~#
```
Third point would be the detail that with docker-compose v3 and higher we'd need to use the docker-swarm options which would require various other updates to documentation if I got that correctly. See [docker-compose doc1](https://docs.docker.com/compose/compose-file/#mem_limit ), [docker-compse/swarm doc2](https://docs.docker.com/compose/compose-file/deploy/#memory) and [here](https://stackoverflow.com/questions/42345235/how-to-specify-memory-cpu-limit-in-docker-compose-version-3).

Most of the memory related stuff is from [here](https://faun.pub/understanding-docker-container-memory-limit-behavior-41add155236c) and [here](https://medium.com/geekculture/docker-limit-container-memory-usage-6ef641533cc2). Careful, second link goes to medium.com.

And of course we have the fact that even if that would do, it would only help with the docker install. For barebone installs this wouldn't help. Not that I'd have any better idea, just wanted to bring up what I found when I looked into that specific idea earlier while debugging my problem ;)

**Comment by Rondal** [comment]
> I checked what I imagine are the relevant tools between 1.6 and 1.7:
> 
> tool	version 1.6	version 1.7
> ocrmypdf	13.4.0	13.4.2
> tesseract	4.1.1	4.1.1
> unpaper	6.1	6.1
> gs (ghostscript)	9.53.3	9.53.3
> pdfminer.six	2021.10.12	2022.03.19
> So basically no version changes, only a minor bump to ocrmypdf, and just 1 version in pdfminer.six. I do find [pdfminer.six/issues/723](https://github.com/pdfminer/pdfminer.six/issues/723), which could be related?

I don't know the content of your document, but looking at the error with my sample documents earlier in this issue I'd say the pdfminer issue is **probably** unrelated. There were some changes that seem not beeing minor on ocrmypdf 13.4.1 and 13.4.2. Maybe it's worth a shot raising an issue over there and referencing this one here?

**Comment by stumpylog** [comment]
I've taken 1.7.0 and downgraded just ocrmypdf to 13.4.0 (matching what 1.6.0 released with).  This also required downgrading pdfminer.six.  (See [the diff](https://github.com/paperless-ngx/paperless-ngx/compare/ngx-1.7.0...feature-memory-testing) if you're really curious).

For those who can reproduce this, could you create a test environment and see if this image has better memory performance?  That will help confirm the issue lies with the upgrade of ocrmypdf and may give them some good information about the error.

`ghcr.io/paperless-ngx/paperless-ngx:feature-memory-testing`

**Comment by panther-sh** [comment]
Here are my test results. The paperless-ngx:feature-memory-testing image doesn't seem to bring any improvement.

System:
Raspberry Pi 4 / 2GB Ram
120GB SSD on USB3.0 Port
Raspberry Pi OS 64 bit
Docker version 20.10.14
Swap Partition size = 100MB
PAPERLESS_TASK_WORKER = 2
PAPERLESS_THREADS_PER_WORKER = 2 

Test results:

Tested with two .pdf files. The first file contains 9 and the second 12 pages (300 DPI).

Paperless Version | 1.6.0 | 1.7.0:feature-memory-testing | 1.7.0
-- | -- | -- | --
OOM issues | no | yes | yes
consume speed | file1: 445 seconds file2: 519 seconds | file consume stoped after about 45 minutes | file consume stoped after about 20 minutes
load average | 2-4,5 | 4-60 | 3-4 and then freezed for 20 minutes


<br class="Apple-interchange-newline">


The only option for me is to extend the swap partition. With a swap size of 2GB, there are no further OMM issues and the consume speed ist similar to 1.6.0. About 1,5GB of the swap is in use.
Probably, this is just a solution for an operating system installed on an SSD. I think the increasing writes will shorten the lifespan of an SD card.

**Comment by Rondal** [comment]
> For those who can reproduce this, could you create a test environment and see if this image has better memory performance? That will help confirm the issue lies with the upgrade of ocrmypdf and may give them some good information about the error.
> 
> `ghcr.io/paperless-ngx/paperless-ngx:feature-memory-testing`

I just tested with this image and the documents I shared way above. Memory consumption doesn't really change compared to 1.7.0 (release) and is still noticeably higher than with 1.6.0. Strange. That would basically rules out the version bumps. 

| tool | version 1.6 | version 1.7 memory|version 1.7 |
|-----|--------------|-------------|----------|
|~~ocrmypdf~~|13.4.0|13.4.0|13.4.2|
|~~tesseract~~|4.1.1|4.1.1|4.1.1|
|~~unpaper~~|6.1|6.1|6.1|
|~~gs (ghostscript)~~|9.53.3|9.53.3|9.53.3|
|~~pdfminer.six~~|2021.10.12|2021.10.12|2022.03.19|

I'll have another look at the other changes from 1.6 to 1.7 but last time I did not notice anything interesting in the way the other tools are called.

Let me know if I can help with any additional data.

**Comment by Limerick-gh** [comment]
I ran into the same issue on Raspberry Pi 4 with 2GB using 1.8.0. Looks like processes were killed resulting in documents not being processed due to high memory usage. Since this issue is already open since quite some months, I'm wondering if any of you who has same issues have found a workaround. 
If someone can confirm that extending the swap partition does help, I would try to put it on USB drive.

**Comment by panther-sh** [comment]
On my Pi4 (2GB Ram / 2GB swap) paperless 1.8.0 (1.7.0) has been running for three months without OOM issues.

**Comment by Limerick-gh** [comment]
Thanks for sharing @panther-sh, much appreciated. I just did the setup with an 4GB USB flash drive (didn't had a smaller one available) for swap and processed a file successfully where it always stopping before. So, fingers crossed. :-)

**Comment by Rondal** [comment]
Hi, original reporter here. I've ignored the initial problem for a while by just increasing the memory to 10GB. Now after upgrading to the most recent release (coming from 1.7) it seems that I also can't reproduce the issues anymore with an otherwise unchanged system. Haven't done much testing in that regard though.

Just wanted to leave that feedback :)

**Comment by shamoon** [comment]
Thanks for the follow-up and everyone here for their contributions.
